# LAB 1 Елементи мови Python. Основні інструкції та структури даних. Робота з файловою системою

Нехай протягом навчання програмні проєкти з різних предметів зберігалися в папках …  .
Є перелік папок <code>folders</code>, в яких зберігаються ці проєкти (задає виконавець, виходячи з вмісту своїх носіїв даних).
<br>
<i>Примітка</i>. До цього переліку мають входити не окремі папки з проєктами, а папки, що містять, наприклад, всі проєкти такої-то дисципліни такого-то семестру тощо. Інакше кажучи, не папка одного solution, а папка, в якій зберігаються solutions (якщо в термінології Visual Studio).

Щоб профільтрувати допоміжні файли середовищ, введемо "білий список" розширень (<code>WHITE_LIST</code>) та список розширень файлів з текстом програм(задається) (<code>CODE_LIST</code>).

З'ясувати

<ol>
    <li>Скільки різних розширень зустрічається у папках з заданого списку (тут і далі: включаючи вкладені), скільки файлів з кожним розширенням.
    (Для цієї задачі розглядаємо всі розширення, а не тільки "білого списку".)</li>    
    Відповідну інформацію вивести на екран охайною табличкою. На рядку виводиться розширення, кількість файлів та ознака (C,W,O відповідно для файлів з текстом програм (відповідно до <code>CODE_LIST</code>), інших з білого списку, та тих, що не потрапили в білий список). Інформація подається за спаданням кількості файлів. 

Наприклад, 
<code>
   .txt     155    W
 .ipynb      73    W
   .cpp      45    C
   .exe       3    O
</code>
    <li>За датами створення файлів для кожного місяця, починаючи з вересня 2024-го року, знайти таку інформацію:
    кількість файлів з розширеннями з "білого списку", їх сумарний обсяг у байтах, кількість їх рядків (враховувати тільки для файлів категорії C; білі та порожні рядки не рахувати), середній розмір файлу, середня кількість рядків.</li>
    Відповідну інформацію вивести на екран охайною табличкою в хронологічному порядку.
    
 Наприклад, 
<code>
 2024  09   23   123456  256   5368   11
 2024  10   23   123456  256   5368   11
 2024  11  134   123456  256    921   11  
</code>      
    <li> Для кожного місяця та кожного розширення, що зустрічаються при перегляді файлів, розміщених у папках зі списку, знайти таку інформацію:
    рік, місяць, розширення, кількість файлів, сумарний обсяг у байтах, кількість їх рядків (тільки для розширень категорії C; білі та порожні рядки не рахувати; для інших розширень цю клітину таблиці залишаємо незаповненою)</li>   
    Цю інформацію записати в csv-файл з іменем <code>lab1_1.csv</code> (кодування тільки utf-8), на початку додати рядок з заголовками: year, mon, ext, cnt, size, lines, avg_size, avg_line.
  
</ol>

Здається:
<ol>
    <li>Цей робочий зошит (заповнений, вилучати клітини, які були в ньому з самого початку, не дозволяється) та pdf-звіт.</li> 
    Перед здачею необхідно зробити Kernel -> Restart Kernel and Run All Cells... Для отримання pdf-звіта блокнот можна просто роздрукувати.
    <br>
    <b>Увага!</b> Повідомлень про помилки виконання бути не повинно.
    <li>Файли зі скриптами для вирішення трьох задач (див. вище).</li>
    <li>Файл <code>lab1_1.csv</code> </li>
</ol>

Оцінювання (максимум 15 балів):

Обов'язково має бути повністю та правильно виконаний (без повідомлень про помилки) та заповнений робочий зошит (блокнот), має бути pdf-звіт та <code>lab1_1.csv</code>, що відповідає виконаному блокноту.

За відсутності скриптів оцінка не може перевищувати 12 балів.

Правильність виконання оцінюється з точки зору дотримання вимог та отриманого кінцевого результату.

Якщо було вибрано папки майже без коду, то оцінка не може перевищувати 6 балів.

За неохайне виведення знімається 3 бали.

In [1]:
CODE_LIST = {'.cpp', '.hpp', '.c', '.h', '.py', '.pyw', '.cs', '.js', '.java'}
WHITE_LIST = CODE_LIST | {'.html', '.json', '.txt', '.ipynb'}

In [2]:
import sys
print(sys.platform, sys.version, sep='\n')

linux
3.14.2 (main, Jan 24 2026, 16:11:19) [GCC 15.2.0]


Інструкції імпортування зазвичай записують на початку. Далі вони будуть додатково продубльовані по місцю використання.

In [3]:
import datetime
from pathlib import Path
from datetime import datetime as dt
import csv

## Вихідні відомості

In [4]:
language = 'python3'

Бібліотека <code>datetime</code> містить стандартні засоби для зображення дат, зокрема клас <a href="https://docs.python.org/3/library/datetime.html#date-objects" target=_blank 
rel="noopener noreferrer">datetime.date</a>

In [5]:
import datetime
year = datetime.date.today().year

In [6]:
year   

2026

In [7]:
f'Ця робота використовує мову програмування {language} і була виконана у {year} році.'

'Ця робота використовує мову програмування python3 і була виконана у 2026 році.'

In [8]:
author = "впишіть власне ім'я та прізвище"
group = "впишіть власну групу"
lab_instructor = "впишіть хто веде лабораторні заняття"

Запишіть f-рядок, значенням якого є текст (розташовується на двох рядках)
<br>
<code>Виконавець: ...., група ...
Викладач:
</code>
Значення виконавця, його групи та викладача мають підставлятися зі змінних <code>author</code>, <code>group</code> та <code>lab_instructor</code>.

<details>
    <summary>Підказка. </summary>
    Доцільно використати рядок у <a href="https://docs.python.org/3/reference/lexical_analysis.html#triple-quoted-strings" target=_blank 
rel="noopener noreferrer">трикратних лапках</a>.
</details>

In [9]:
info = None     # тут має бути записаний питомий f-рядок
print(info)

None


## Готуємось "гуляти" по папках (підготовка-дослідження)

Нам знадобиться клас <code>Path</code> з бібліотеки <code>pathlib</code>. <a href="https://docs.python.org/3/library/pathlib.html" target=_blank 
rel="noopener noreferrer">Документація</a>.

In [10]:
from pathlib import Path

Дізнаємось поточний робочий каталог. (<b>c</b>urrent <b>w</b>ork <b>d</b>ir)

In [11]:
Path().cwd()

PosixPath('/media/tania/Apacer PHD/gits/Courses/Py2/code')

In [12]:
str(Path().cwd())

'/media/tania/Apacer PHD/gits/Courses/Py2/code'

Шлях до якої-небудь папки з помірною кількістю файлів з різними розширеннями та складеними папками зробити значенням змінної <code>root</code>.
<br>
У кожного має бути свій шлях.

Наприклад, можна взяти папку з проєктом мови C++.

In [13]:
root = Path('/media/tania/Apacer PHD/gits/Courses/C++1/projects/lab07')
root

PosixPath('/media/tania/Apacer PHD/gits/Courses/C++1/projects/lab07')

### Обійти папку, вивести всі наявні в ній файли

Цікавить шлях, тип файлу (каталог чи звичайний файл), для звичайного файлу цікавить розширення та розмір у байтах.

Сімлінки та інше будемо пропускати.

<details> 
    <summary>
        Підказка 1
    </summary>
    У класі є метод <code>glob</code>, що здатен вирішити задачу. Див.  
    <a href="https://docs.python.org/3/library/pathlib.html" target=_blank 
rel="noopener noreferrer">документацію</a> бібліотеки. 
</details>

In [14]:
help(Path.glob)

Help on function glob in module pathlib:

glob(self, pattern, *, case_sensitive=None, recurse_symlinks=False)
    Iterate over this subtree and yield all existing files (of any
    kind, including directories) matching the given relative pattern.



<details> 
    <summary>
        Підказка 2
    </summary>
    Для рекурсивного обходу каталогу можна використати шаблон <code>**/*</code>, за яким будуть обходитись усі елементи, крім самої папки, що обходиться ( . ). 
</details>

<details> 
    <summary>
        Підказка 3
    </summary>
    Якщо <code>s</code> задає послідовність або те, по чому можна ітерувати, то обхід можна реалізувати за допомогою циклу  <code>for</code>

<code>for el in s:
    do something with s
</code>

</details>

Обійти всі наявні в <code>root</code> файли та вивести їх імена

In [15]:
# заповнити

/media/tania/Apacer PHD/gits/Courses/C++1/projects/lab07/build
/media/tania/Apacer PHD/gits/Courses/C++1/projects/lab07/CMakeLists.txt
/media/tania/Apacer PHD/gits/Courses/C++1/projects/lab07/lab7.cpp
/media/tania/Apacer PHD/gits/Courses/C++1/projects/lab07/lab7a.cpp
/media/tania/Apacer PHD/gits/Courses/C++1/projects/lab07/lab7a.h
/media/tania/Apacer PHD/gits/Courses/C++1/projects/lab07/lab7b.cpp
/media/tania/Apacer PHD/gits/Courses/C++1/projects/lab07/lab7b.h
/media/tania/Apacer PHD/gits/Courses/C++1/projects/lab07/build/.cmake
/media/tania/Apacer PHD/gits/Courses/C++1/projects/lab07/build/CMakeCache.txt
/media/tania/Apacer PHD/gits/Courses/C++1/projects/lab07/build/CMakeFiles
/media/tania/Apacer PHD/gits/Courses/C++1/projects/lab07/build/cmake_install.cmake
/media/tania/Apacer PHD/gits/Courses/C++1/projects/lab07/build/compile_commands.json
/media/tania/Apacer PHD/gits/Courses/C++1/projects/lab07/build/lab07
/media/tania/Apacer PHD/gits/Courses/C++1/projects/lab07/build/Makefile
/med

Є бажання виводити їх шляхи відносно заданої папки.

Тобто необхідно побудувати шлях, відносно шляху <code>root</code>. Допоможе метод <code>relative_to</code>.

In [16]:
help(Path.relative_to)

Help on function relative_to in module pathlib:

relative_to(self, other, *, walk_up=False)
    Return the relative path to another path identified by the passed
    arguments.  If the operation is not possible (because this is not
    related to the other path), raise ValueError.

    The *walk_up* parameter controls whether `..` may be used to resolve
    the path.



<i>Примітка</i>. Насправді це метод класу <code>PurePath</code>, який є базовим для <code>Path</code>.

Після завершення циклу змінна <code>el</code>, яка позначала поточний елемент обходу, продовжує існувати.
Проведемо з нею кілька експериментів.

Виведемо її значення.

In [17]:
el

PosixPath('/media/tania/Apacer PHD/gits/Courses/C++1/projects/lab07/build/.cmake/api/v1/query/client-vscode/query.json')

У наступній клітині побудуємо та виведемо відповідний відносний шлях.
<details> 
    <summary>
        Підказка
    </summary>
    <code>el.relative_to(root)</code>
</details>

PosixPath('build/.cmake/api/v1/query/client-vscode/query.json')

Явно перетворимо відносний шлях на рядок (значення типу <code>str</code>).

<details> 
    <summary>
        Підказка
    </summary>
    Щоб об'єкт <code>obj</code> перетворити на об'єкт типу <code>T</code>, можна використати конструктор типу <code>T</code>, а саме <code>T(obj)</code>.
</details>

<br>
<details> 
    <summary>
        Відповідь
    </summary>
<code>str(el.relative_to(root))</code>
</details>

In [19]:
# заповнити

'build/.cmake/api/v1/query/client-vscode/query.json'

Тепер у нашому обході будемо виводити шляхи, відносно шляху <code>root</code>.
У наступну клітину слід записати модифікацію обходу, який здійснювався раніше.

<i>Примітка</i>. Якщо ми збираємось відносний шлях тільки виводити, то перед виведенням перетворювати його явно на рядок потреби немає. Це чудово зробить сама функція <code>print</code>.

<details> 
    <summary>
        Підказка
    </summary>
<code>for el in <i>вираз, що задає обхід</i>:
    print(відносний шлях)
</code>
</details>

In [20]:
# заповнити

build
CMakeLists.txt
lab7.cpp
lab7a.cpp
lab7a.h
lab7b.cpp
lab7b.h
build/.cmake
build/CMakeCache.txt
build/CMakeFiles
build/cmake_install.cmake
build/compile_commands.json
build/lab07
build/Makefile
build/.cmake/api
build/CMakeFiles/4.1.2
build/CMakeFiles/cmake.check_cache
build/CMakeFiles/CMakeConfigureLog.yaml
build/CMakeFiles/CMakeDirectoryInformation.cmake
build/CMakeFiles/CMakeScratch
build/CMakeFiles/InstallScripts.json
build/CMakeFiles/lab07.dir
build/CMakeFiles/Makefile.cmake
build/CMakeFiles/Makefile2
build/CMakeFiles/pkgRedirects
build/CMakeFiles/progress.marks
build/CMakeFiles/TargetDirectories.txt
build/CMakeFiles/4.1.2/CMakeCCompiler.cmake
build/CMakeFiles/4.1.2/CMakeCXXCompiler.cmake
build/CMakeFiles/4.1.2/CMakeDetermineCompilerABI_C.bin
build/CMakeFiles/4.1.2/CMakeDetermineCompilerABI_CXX.bin
build/CMakeFiles/4.1.2/CMakeSystem.cmake
build/CMakeFiles/4.1.2/CompilerIdC
build/CMakeFiles/4.1.2/CompilerIdCXX
build/CMakeFiles/lab07.dir/build.make
build/CMakeFiles/lab07.dir/cmak

### Складові шляху

Інформацію про частини шляху можна знайти <a href="https://docs.python.org/3/library/pathlib.html#methods-and-properties" target=_blank 
rel="noopener noreferrer">тут</a>.

Альтернатива: використати довідку.

Знайдемо всі атрибути, доступні для класу <code>Path</code>. (У Python під атрибутами розуміють і класичні атрибути (описують дані об'єкта), і методи (це наслідок технічної реалізації методів).)

Функція <code>dir</code> поверне список.

In [21]:
dir(Path)

['__bytes__',
 '__class__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__firstlineno__',
 '__format__',
 '__fspath__',
 '__ge__',
 '__getattribute__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__rtruediv__',
 '__setattr__',
 '__sizeof__',
 '__slots__',
 '__static_attributes__',
 '__str__',
 '__subclasshook__',
 '__truediv__',
 '_copy_from',
 '_copy_from_file',
 '_copy_from_symlink',
 '_delete',
 '_drv',
 '_filter_trailing_slash',
 '_format_parsed_parts',
 '_from_dir_entry',
 '_from_parsed_parts',
 '_from_parsed_string',
 '_hash',
 '_info',
 '_parse_path',
 '_parse_pattern',
 '_parts_normcase',
 '_parts_normcase_cached',
 '_raw_path',
 '_raw_paths',
 '_remove_leading_dot',
 '_remove_trailing_slash',
 '_root',
 '_str',
 '_str_normcase',
 '_str_normcase_cached',
 '_tail',
 '_tail_cached',
 'absolute',
 'anchor',
 'as_posix',
 'as_uri',
 'c

Серед них ті, що починаються з двох підкреслень, є службовими. Ті, що починаються з одного підкреслення, &mdash;   допоміжними (не є відкритими). Виведемо всі, крім них.

In [22]:
cnt = 0
for a in dir(Path):
    if not a.startswith('_'):
        print(a, end=', ')
        cnt += 1
        if (cnt > 5):
            cnt = 0
            print()

absolute, anchor, as_posix, as_uri, chmod, copy, 
copy_into, cwd, drive, exists, expanduser, from_uri, 
full_match, glob, group, hardlink_to, home, info, 
is_absolute, is_block_device, is_char_device, is_dir, is_fifo, is_file, 
is_junction, is_mount, is_relative_to, is_reserved, is_socket, is_symlink, 
iterdir, joinpath, lchmod, lstat, match, mkdir, 
move, move_into, name, open, owner, parent, 
parents, parser, parts, read_bytes, read_text, readlink, 
relative_to, rename, replace, resolve, rglob, rmdir, 
root, samefile, stat, stem, suffix, suffixes, 
symlink_to, touch, unlink, walk, with_name, with_segments, 
with_stem, with_suffix, write_bytes, write_text, 

Повертаємось до нашого примірника шляху.

In [23]:
el

PosixPath('/media/tania/Apacer PHD/gits/Courses/C++1/projects/lab07/build/.cmake/api/v1/query/client-vscode/query.json')

Визначимо, чи задає цей шлях папку, файл, яке в нього ім'я, розширення, який його розмір.

<details>
<summary>Підказка</summary>
    Використати атрибути <code>is_dir</code>, <code>is_file</code>, <code>name</code>, <code>suffix</code>, <code>stat().st_size</code>
</details>

Записати вираз, що перевіряє, чи задає шлях <code>el</code> папку.

In [24]:
# заповнити

False

Записати вираз, що перевіряє, чи задає шлях <code>el</code> файл.

In [25]:
# заповнити

True

Записати вираз, що визначає ім'я файлу <code>el</code>.

In [26]:
# заповнити

'query.json'

Записати вираз, що визначає розширення файлу <code>el</code>.

In [27]:
# заповнити

'.json'

Записати вираз, що визначає розмір файлу <code>el</code>.

In [28]:
# заповнити

144

Для розв'язання загальної задачі нам буде необхідно вміти визначати рік та місяць створення файла.

Не тільки розмір файлу, але й інша інформація про файл може бути знайдена за допомогою метода <code>stat</code>
(<a href="https://docs.python.org/3/library/pathlib.html#pathlib.Path.stat" target=_blank 
rel="noopener noreferrer">документація</a>)

Нас цікавить дата створення файла. Тут виникає невелика залежність від ОС. (<a href="https://docs.python.org/3/library/os.html#os.stat_result" target=_blank 
rel="noopener noreferrer">документація</a>)

Знайдемо час створення файлу (у секундах) і запам'ятаємо в змінній <code>ctime</code>.

In [29]:
if sys.platform.startswith('win'):
    ctime = el.stat().st_birthtime
else:
    ctime = el.stat().st_ctime
ctime    

1761171895.1253946

Тепер слід перетворити на звичайну дату, щоб з'ясувати рік та місяць.

Використаємо клас <code>datetime</code>, точніше його метод <code>fromtimestamp</code>. (<a href="https://docs.python.org/3/library/datetime.html#datetime.datetime.fromtimestamp" target=_blank 
rel="noopener noreferrer">документація</a>)

Нижче клас <code>datetime</code> імпортований з одноіменної бібліотеки <code>datetime</code> під псевдонімом <code>dt</code>.

In [30]:
from datetime import datetime as dt

In [31]:
cdate = dt.fromtimestamp(ctime)
cdate

datetime.datetime(2025, 10, 23, 1, 24, 55, 125395)

In [32]:
cdate.year, cdate.month

(2025, 10)

Напишіть <b>кросплатформену</b> функцію <code>created_at</code>, що за шляхом до файлу визначає рік та місяць його створення. Функція має отримувати шлях до файлу та повертати пару цілих чисел.

Приклад функції, що повертає два значення.
<code>
def f(arg):
    return arg, arg+arg
</code>

<b>Питання для обговорення</b>. В який момент краще розгалужити обчислення дати створення?

<details>
<summary>Відповідь</summary>
<code>
def _created_at_Win(...): ...

def _created_at_Linux(...): ...

if sys.platform.startswith('win'):
    created_at = _created_at_Win
else:
    created_at = _created_at_Linux
</code>
</details>

In [33]:
def _created_at_Linux(path):     # треба записати власні реалізації цих двох функцій
    throw NotImplemented

def _created_at_Windows(path):
    throw NotImplemented

if sys.platform.startswith('win'):
    created_at = _created_at_Win
else:
    created_at = _created_at_Linux    

Перевіримо працездатність функції.

In [34]:
created_at(el)

(2025, 10)

In [35]:
y, m = created_at(el)
print(y, m)

2025 10


Повертаємось до задачі обходу та виведення

Тепер для шляху <code>el</code> виведемо:
<br>
відповідний йому відносний шлях, далі на тому ж рядку <code>file</code> або <code>dir</code>
<br>
для файлу на новому рядку з абзацного відступу виведемо його розширення в 10 символів та розмір в 10 символів

In [36]:
s = '.txt'
size = 12345
print(f'\t {s:10} {size:10}')

	 .txt            12345


Записати код, що виводить інформацію про файл як у прикладі нижче

In [37]:
# заповнити

build/.cmake/api/v1/query/client-vscode/query.json --- file
	.json      144


<i>Приклад формату виведення</i>:
<code>
build/.cmake/api/v1/query/client-vscode/query.jsonfile --- file
	.json      144
</code>

Зберемо обхід, за якого будемо виводити інформацію тільки про папки та звичайні файли.

У базовий обхід додаємо виведення інформації про файли.

In [38]:
# заповнити

build --- dir
CMakeLists.txt --- file
	.txt       411
lab7.cpp --- file
	.cpp       461
lab7a.cpp --- file
	.cpp       1032
lab7a.h --- file
	.h         53
lab7b.cpp --- file
	.cpp       1337
lab7b.h --- file
	.h         51
build/.cmake --- dir
build/CMakeCache.txt --- file
	.txt       14691
build/CMakeFiles --- dir
build/cmake_install.cmake --- file
	.cmake     2187
build/compile_commands.json --- file
	.json      1168
build/lab07 --- file
	           647448
build/Makefile --- file
	           6642
build/.cmake/api --- dir
build/CMakeFiles/4.1.2 --- dir
build/CMakeFiles/cmake.check_cache --- file
	.check_cache 85
build/CMakeFiles/CMakeConfigureLog.yaml --- file
	.yaml      94573
build/CMakeFiles/CMakeDirectoryInformation.cmake --- file
	.cmake     693
build/CMakeFiles/CMakeScratch --- dir
build/CMakeFiles/InstallScripts.json --- file
	.json      139
build/CMakeFiles/lab07.dir --- dir
build/CMakeFiles/Makefile.cmake --- file
	.cmake     10303
build/CMakeFiles/Makefile2 --- file
	      

Визначимо, файли з якими розширеннями присутні в папці, що обходиться.

<details>
    <summary>Підказка 1 (про алгоритм).</summary>
Створимо порожню множину. У процесі обходу папки будемо додавати в неї знайдені розширення. Наприкінці виведемо знайдені розширення в алфавітному порядку.        
</details>

<br>
<details>
    <summary>Підказка 2 (про деталі реалізації).</summary>

<ul>
<li>Типом множин є <code>set</code> (<a href="https://docs.python.org/3/library/stdtypes.html#set-types-set-frozenset" target=_blank 
rel="noopener noreferrer">документація</a>).</li>
   
<li> Щоб додати елемент до множини, можна використати метод <code>add</code> (<a href="https://docs.python.org/3/library/stdtypes.html#set.add" target=_blank 
rel="noopener noreferrer">документація</a>).</li>    
<li>Відсортувати допоможе функція <code>sorted</code>(<a href="https://docs.python.org/3/library/functions.html#sorted" target=_blank 
rel="noopener noreferrer">документація</a>). Зверніть увагу на те, як задати сортування у спадному порядку.</li>    
</ul>
</details>

Значенням змінної <code>suffixes</code> зробіть множину наявних розширень файлів. 

In [39]:
# заповнити

In [40]:
suffixes

{'',
 '.bin',
 '.c',
 '.check_cache',
 '.cmake',
 '.cpp',
 '.d',
 '.h',
 '.json',
 '.make',
 '.marks',
 '.o',
 '.out',
 '.ts',
 '.txt',
 '.yaml'}

Перетворіть її на відсортований список <code>suffixes_sorted</code>.

In [41]:
suffixes_sorted = # заповнити

Виведіть знайдені розширення у відсортованому порядку.

In [42]:
for el in suffixes_sorted:
    print(el, end=', ')
print()

, .bin, .c, .check_cache, .cmake, .cpp, .d, .h, .json, .make, .marks, .o, .out, .ts, .txt, .yaml, 


Скільки різних розширень було знайдено? (Виведіть це значення. Вручну нічого не рахуємо!)

<details>
<summary>Підказка</summary>
    Функція <code>len</code> допоможе знайти кількість елементів довільного контейнера.
</details>

In [43]:
# заповнити

16

Знайдемо скільки файлів з певним розширенням наявно у досліджуваній папці.

Для кожного розширення визначимо, скільки файлів з цим розширенням присутні в папці, що обходиться.

<details>
    <summary>Підказка 1 (про алгоритм).</summary>
Створимо словник, ключами в якому будуть розширення файлів. Значеннями словника будуть лічильники кількості файлів з відповідним розширенням. У процесі обходу папки будемо збільшувати лічильники.      
</details>

<br>
<details>
    <summary>Підказка 2 (про деталі реалізації).</summary>

<ul>
<li>Типом словників є <code>dict</code> (<a href="https://docs.python.org/3/library/stdtypes.html#mapping-types-dict" target=_blank 
rel="noopener noreferrer">документація</a>).</li>
   
<li> Наявність ключа в словнику можна перевірити за допомогою оператора <code>in</code>. Або у нагоді може стати метод <code>setdefault</code> .</li>    
  
</ul>
</details>

Значенням змінної <code>d</code> зробіть словник, що знає кількість файлів для кожного наявного розширення. 

In [44]:
# заповнити

In [45]:
d

{'.txt': 4,
 '.cpp': 4,
 '.h': 2,
 '.cmake': 8,
 '.json': 10,
 '': 3,
 '.check_cache': 1,
 '.yaml': 1,
 '.marks': 1,
 '.bin': 2,
 '.make': 5,
 '.ts': 1,
 '.o': 3,
 '.d': 3,
 '.out': 2,
 '.c': 1}

In [46]:
suffixes_cnts = list(d.items())
for el in suffixes_cnts:
    print(el, end=', ')
print()

('.txt', 4), ('.cpp', 4), ('.h', 2), ('.cmake', 8), ('.json', 10), ('', 3), ('.check_cache', 1), ('.yaml', 1), ('.marks', 1), ('.bin', 2), ('.make', 5), ('.ts', 1), ('.o', 3), ('.d', 3), ('.out', 2), ('.c', 1), 


Знайдемо найбільш популярні розширення.

<details> 
    <summary>Підказки</summary>

<ol>
    <li>Серед <a href="https://docs.python.org/3/library/functions.html" target=_blank 
rel="noopener noreferrer">вбудованих функцій</a> Python зверніть увагу на <code>max</code> (<a href="https://docs.python.org/3/library/functions.html#max" target=_blank 
rel="noopener noreferrer">документація</a>). </li>
    <li> <a href="https://docs.python.org/3/reference/expressions.html#lambda" target=_blank 
rel="noopener noreferrer">Анонімні функції </a> теж будуть у нагоді. </li>
</ol>
    
</details>

Є список пар.

In [47]:
lst = [(1, 2), (1, 5), (8, 2), (2, 4)]
lst

[(1, 2), (1, 5), (8, 2), (2, 4)]

Відсортуємо пари лексикографічно.

In [48]:
lst.sort()
lst

[(1, 2), (1, 5), (2, 4), (8, 2)]

Відсортуємо пари за ключем. В якості ключа візьмемо обернену пару.

In [49]:
lst.sort(key = lambda el: (el[1], el[0]))
lst

[(1, 2), (8, 2), (2, 4), (1, 5)]

Знайти найбільшу кількість файлів, що мають однакове розширення. Присвоїти значення змінній <code>max_count</code>.

<details><summary>Підказка</summary>
Ключем має бути другий елемент пари.</details>

In [50]:
max_count = # заповнити
max_count

('.json', 10)

In [51]:
max_count = max_count[1]
max_count

10

Вивести всі розширення, що відповідають цій найбільшій кількості входжень.

<details>
    <summary>Підказка</summary>
    Як варіант: обійти пари <code>.items()</code>, що зберігаються в словнику, і вивести потрібні розширення.
</details>

In [52]:
# заповнити

.json


##### Класифікація розширень

Як за розширенням визначити, чи потрапляє воно в задані переліки?

<details>
    <summary>Підказка</summary>
    Використайте оператор <code>in</in>
</details>

Запишіть логічний вираз, що задає ознаку того, що розширення .txt наявне у білому списку.

<details>
    <summary>Відповідь</summary>
<code> 
'.txt' in WHITE_LIST
</code>
</details>

In [53]:
# заповнити

True

Напишіть функцію <code>kind</code>, що за розширенням повертає рядок з символом W,C,O (див. умову)

In [54]:
def kind(extension):   # записати
    path

Модифікуйте словник так, щоб для кожного розширення він зберігав:
<br>
<ol>
    <li>кількість файлів</li>
    <li>їх сумарний розмір</li>
    <li>їх сумарну кількість рядків тексту (тільки для файлів з текстом програм, для інших &mdash; $0$)</li>
</ol>

Значеннями словника можуть бути списки з трьох елементів (кількість файлів, сумарний розмір, сумарна кількість рядків тексту)

In [55]:
sample = {'.cpp':[23, 123456, 56]}   # приклад

При перегляді файлів слід оновлювати поля значення. Наприклад
<br>
<code>sample['.cpp'][0]+=1</code>    збільшує кількість файлів з розширенням на 1
<br>
<code>sample['.cpp'][1]+=1234</code>  додає розмір файлу (1234) до сумарного розміру


Якщо треба рахувати інформацію не тільки в розрізі розширень, а ще й місяців, то в якості ключів можна використати трійки: рік, місяць, розширення.

In [56]:
sample = {(2024, 9, '.cpp'):[23, 123456, 56]}   # приклад
sample

{(2024, 9, '.cpp'): [23, 123456, 56]}

In [57]:
sample[2024, 9, '.cpp']

[23, 123456, 56]

##### Кількість рядків файлу

Створити шлях до якого-небудь власного файлу з текстом програми. (Підставити власний шлях.)

In [58]:
sample = root.joinpath('lab7.cpp')

In [59]:
sample.stat().st_size

461

Знайдемо кількість рядків файлу

In [60]:
if sample.stat().st_size < 2**20:
    lines = 0
    with open(sample) as f:
        for _ in f:  # ітеруємо рядками    _  тут ім'я змінної, що обходить рядки файлу, можна змінити
            lines += 1
    print(lines)

28


Якщо рядок порожній, то він хибний. Якщо рядок містить тільки білі символи, то це можна визначити за допомогою методу <code>isspace</code> (<a href="https://docs.python.org/3/library/stdtypes.html#string-methods" target=_blank 
rel="noopener noreferrer">методи класу <code>str</code></a>)

In [61]:
help(str.isspace)

Help on method descriptor isspace:

isspace(self, /) unbound builtins.str method
    Return True if the string is a whitespace string, False otherwise.

    A string is whitespace if all characters in the string are whitespace and there
    is at least one character in the string.



Модифікувати підрахунок кількості рядків файлу так, щоб додатково рахувалась кількість рядків файлу, що містять небілі символи (непорожні і не тільки з білих символів). У кінці вивести кількість рядків та кількість рядків з небілими символами.

In [62]:
if sample.stat().st_size < 2**20:
    lines = 0
    notempty_lines = 0
    with open(sample) as f:
        for s in f:  # ітеруємо рядками
            lines += 1
            if # записати умову:
                notempty_lines += 1
    print(lines, notempty_lines)

28 26


Написати функцію, що визначає для файла кількість його рядків з небілими символами.

<code>
def lines_counter(path):
    ???
    return ???
</code>

In [63]:
def lines_counter(path):
    # належно написати
    return 26

Перевіримо нашу фукцію зоча б трішечки.

In [64]:
lines_counter(sample)

26

### Як зберегти табличну інформацію в csv-файл

Приклад даних. Табличні дані (як список контейнерів). (Використовувати списки чи кортежі тут різниці не має.) Та перелік заголовків до таблиці.

In [65]:
data = [(2025, 1, 13, 23456, 4.5, 99, 'текст1'), 
        (2025, 7, 35, 23456, 4.5, 99, 'текст2'), 
        (2025, 4, 34, 23456, 4.5, 99, 'текст3'), 
        (2025, 11, 24, 23456, 4.5, 99, 'текст1')]
headers = ['h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'h7']

Роботу виконуватиме бібліотека <csv> (<a href="https://docs.python.org/3/library/csv.html" target=_blank 
rel="noopener noreferrer">документація</a>), більш конкретно <code>writer</code> (<a href="https://docs.python.org/3/library/csv.html#csv.writer" target=_blank 
rel="noopener noreferrer">документація</a>) . Для читання є <code>reader</code> (<a href="https://docs.python.org/3/library/csv.html#csv.reader" target=_blank 
rel="noopener noreferrer">документація</a>).

In [66]:
import csv

In [67]:
with open('data.csv', 'w', encoding='utf-8', newline='') as f:
    writer = csv.writer(f)
    writer.writerow(headers)
    for row in data:
        writer.writerow(row)

In [68]:
with open('data.csv', encoding='utf-8', newline='') as f:
    reader = csv.reader(f)
    headers = next(reader)
    lst = list()
    for row in reader:
        lst.append(row)
        print(row)

['2025', '1', '13', '23456', '4.5', '99', 'текст1']
['2025', '7', '35', '23456', '4.5', '99', 'текст2']
['2025', '4', '34', '23456', '4.5', '99', 'текст3']
['2025', '11', '24', '23456', '4.5', '99', 'текст1']


## Збирання задачі

Нехай протягом навчання програмні проєкти з різних предметів зберігалися в папках …  .
Є перелік папок <code>folders</code>, в яких зберігаються ці проєкти (задає виконавець, виходячи з вмісту своїх носіїв даних).
<br>
<i>Примітка</i>. До цього переліку мають входити не окремі папки з проєктами, а папки, що містять, наприклад, всі проєкти такої-то дисципліни такого-то семестру тощо. Інакше кажучи, не папка одного solution, а папка, в якій зберігаються solutions (якщо в термінології Visual Studio).

In [1]:
folders = []   # задати перелік
folders

[]

Щоб профільтрувати допоміжні файли середовищ, введемо "білий список" розширень (<code>WHITE_LIST</code>) та список розширень файлів з текстом програм(задається) (<code>CODE_LIST</code>).

Ці переліки були задані в умові.

In [69]:
for el in WHITE_LIST: print(el, end=' ')

.html .ipynb .cpp .java .cs .h .js .py .txt .pyw .c .hpp .json 

In [70]:
for el in CODE_LIST: print(el, end=' ')

.cpp .java .cs .h .js .py .pyw .c .hpp 

<b>Задача 1</b> Скільки різних розширень зустрічається у папках з заданого списку (тут і далі: включаючи вкладені), скільки файлів з кожним розширенням.
<br>(Для цієї задачі розглядаємо всі розширення, а не тільки "білого списку".)
    
Відповідну інформацію вивести на екран охайною табличкою. На рядку виводиться розширення, кількість файлів та ознака (C,W,O відповідно для файлів з текстом програм (відповідно до <code>CODE_LIST</code>), інших з білого списку, та тих, що не потрапили в білий список). Інформація подається за спаданням кількості файлів. 

Наприклад, 
<code>
   .txt     155    W
 .ipynb      73    W
   .cpp      45    C
   .exe       3    O
</code>

<ol>
    <li>Створити порожній словник <code>d</code>. Його ключами будуть розширення, а значеннями відповідні кількості файлів.</li>
    <li>Обійти всі папки з переліку <code>folders</code>. Для кожної папки запустити рекурсивний обхід, під час якого рахувати кількості файлів окремо для кожного розширення.</li>
    <li>Перетворити словник на список пар і відсортувати цей список по другій координаті в спадному порядку.</li>
    <li>Обійти список пар і для кожного елемента вивести запитувану інформацію за допомогою f-рядка (налаштовуючи ширину поля виведення(, щоб досягти охайної таблиці.
</ol>

Використати раніше розроблену функцію <code>kind</code>

<b>Задача 2</b> За датами створення файлів для кожного місяця, починаючи з вересня 2024-го року, знайти таку інформацію:
    кількість файлів з розширеннями з "білого списку", їх сумарний обсяг у байтах, кількість їх рядків (враховувати тільки для файлів категорії C; білі та порожні рядки не рахувати), середній розмір файлу, середня кількість рядків.
    
Відповідну інформацію вивести на екран охайною табличкою в хронологічному порядку.
    
 Наприклад, 
<code>
 2024  09   23   123456  256   5368   11
 2024  10   23   123456  256   5368   11
 2024  11  134   123456  256    921   11  
</code>     

<ol>
    <li>Створити порожній словник <code>d</code>. Його ключами будуть пари (рік, місяць), а значеннями відповідні кількості файлів, їх сумарний обсяг, кількість їх рядків.</li>
    <li>Обійти всі папки з переліку <code>folders</code>. Для кожної папки запустити рекурсивний обхід, під час якого обробляти тільки файли з розширеннями з білого списку. При обробленні файлу визначати рік та місяць його створення та оновлювати інформацію в словнику (за потреби додавати ключі).</li>
    <li>Перетворити словник на список пар і відсортувати цей список за хронологією.</li>
    <li>Обійти список пар і для кожного елемента вивести запитувану інформацію за допомогою f-рядка (налаштовуючи ширину поля виведення), щоб досягти охайної таблиці. Середні значення обчислювати під час обходу, не зберігаючи в структурі даних.
</ol>

Використати раніше розроблені раніше функції <code>created_at</code>, <code>lines_counter</code>

<b>Задача 3</b> Для кожного місяця та кожного розширення, що зустрічаються при перегляді файлів, розміщених у папках зі списку, знайти таку інформацію:
    рік, місяць, розширення, кількість файлів, сумарний обсяг у байтах, кількість їх рядків (тільки для розширень категорії C; білі та порожні рядки не рахувати; для інших розширень цю клітину таблиці залишаємо незаповненою)

Цю інформацію записати в csv-файл з іменем <code>lab1_1.csv</code> (кодування тільки utf-8), на початку додати рядок з заголовками: year, mon, ext, cnt, size, lines, avg_size, avg_line.
  

<ol>
    <li>Створити порожній словник <code>d</code>. Його ключами будуть пари (рік, місяць, розширення), а значеннями відповідні кількості файлів, їх сумарний обсяг, кількість їх рядків.</li>
    <li>Обійти всі папки з переліку <code>folders</code>. Для кожної папки запустити рекурсивний обхід, під час якого обробляти тільки файли з розширеннями з білого списку. При обробленні файлу визначати рік та місяць його створення та оновлювати інформацію в словнику (за потреби додавати ключі).</li>
    <li>Перетворити словник на список пар і відсортувати цей список за хронологією.</li>
    <li>Обійти список пар і для кожного елемента вивести запитувану інформацію за допомогою f-рядка (налаштовуючи ширину поля виведення), щоб досягти охайної таблиці. Середні значення обчислювати під час обходу, не зберігаючи в структурі даних.
</ol>

Використати раніше розроблені раніше функції <code>created_at</code>, <code>lines_counter</code>

<i>Примітка</i>. Не забороняється спочатку виконати задачу 3, а потім за її даними отримати результат задачі 2. Але це може виявитись зробити складніше, ніж у пропонованій вище схемі.